### PR TITLE
Add dbt Cloud provider

### PR DESCRIPTION
**Pull request by josh-fell**
This PR adds a new provider to interface with [dbt Cloud](https://docs.getdbt.com/docs/dbt-cloud/cloud-overview) via the [dbt Cloud API](https://docs.getdbt.com/dbt-cloud/api-v2).  The provider includes:

- `DbtCloudHook` which implements an abstraction for almost all of the available endpoints in the dbt Cloud API and inherits from the existing `HttpHook`.
- Two operators, `DbtCloudRunJobOperator` and `DbtCloudGetJobRunArtifactOperator`, which triggers a dbt Cloud job and downloads a run artifact, respectively.
- A sensor, `DbtCloudJobRunSensor`, to poll status of a specific dbt Cloud job run.
- An operator link to allow users to navigate directly to the triggered dbt Cloud job run for monitoring.
- A `test_connection()` method in the `DbtCloudHook` for users to test connections prior to executing DAGs.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/dbt/cloud/operators/dbt.py
Comment (mik-laj) [inline airflow/providers/dbt/cloud/operators/dbt.py]: We should avoid initializing hook in ctor.

File: docs/apache-airflow-providers-dbt-cloud/connections.rst
Comment (mik-laj) [inline docs/apache-airflow-providers-dbt-cloud/connections.rst:66]: Can you add exaample URI as bash commnd?
For exmaple - Google URL
```
export AIRFLOW_CONN_GOOGLE_CLOUD_SQL_DEFAULT='gcpcloudsql://user:XXXXXXXXX@1.1.1.1:3306/mydb?database_type=mysql&project_id=example-project&location=europe-west1&instance=testinstance&use_proxy=True&sql_proxy_use_tcp=False'
```
https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp_sql.html

Comment (josh-fell) [reply docs/apache-airflow-providers-dbt-cloud/connections.rst:66]: Absolutely. Will add.

File: airflow/providers/dbt/cloud/operators/dbt.py
Comment (josh-fell) [reply airflow/providers/dbt/cloud/operators/dbt.py]: Right on. I'll pull this initialization out.

File: docs/apache-airflow-providers-dbt-cloud/connections.rst
Comment (mik-laj) [inline docs/apache-airflow-providers-dbt-cloud/connections.rst]:
```suggestion
        export AIRFLOW_CONN_DBT_CLOUD_DEFAULT='dbt-cloud://:api_token@'
```

File: docs/apache-airflow-providers-dbt-cloud/connections.rst
Comment (mik-laj) [inline docs/apache-airflow-providers-dbt-cloud/connections.rst]: Please use meaningful link text. See: https://developers.google.com/style/cross-references

Comment (josh-fell) [reply docs/apache-airflow-providers-dbt-cloud/connections.rst]: Oh thanks for the tip on this reference. Really nice documentation on best practices.

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (subkanthi) [inline airflow/providers/dbt/cloud/hooks/dbt.py:175]: why is this check useful, it doesnt look like `resp_json['data']` is accessed below.

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (subkanthi) [inline airflow/providers/dbt/cloud/hooks/dbt.py:372]: Should we log an error if account_id is not passed, looks like it might construct the url with None.

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (subkanthi) [inline airflow/providers/dbt/cloud/hooks/dbt.py]: would it help to have a log  statement in the "else" if data is not present in the response.

**Comment by mik-laj** [comment]
@nathaniel-may @emmyoop @leahwicz Can you look at it?

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (josh-fell) [reply airflow/providers/dbt/cloud/hooks/dbt.py:372]: This should be caught by the `@fallback_to_default_account` decorator which wraps every hook method that calls a dbt Cloud API endpoint directly. The decorator checks to see if the `account_id` is not passed or if it is, if the value is `None`. If either of those conditions are met, the `account_id` set in the Airflow connection is used. In the event there is no `account_id` configured in the connection, this exception is raised: `raise AirflowException("Could not determine the dbt Cloud account.")`. Does that suffice?

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (josh-fell) [reply airflow/providers/dbt/cloud/hooks/dbt.py]: Certainly couldn't hurt. Perhaps this is better as a debug statement though since it's more related to the aspects of the API response rather than the task execution.

Comment (josh-fell) [reply airflow/providers/dbt/cloud/hooks/dbt.py]: I removed this check. The `data` object has no bearing on pagination. Thanks for the catch!

**Comment by sungchun12** [comment]
> @nathaniel-may @emmyoop @leahwicz Can you look at it?

@mik-laj These people are dbt-core contributors and do not actively work on the dbt Cloud API.

What did you want someone from dbt Labs to look into specifically for this pull request?

**Comment by mik-laj** [comment]
@sungchun12  DBT is used quite often with Apache Airflow, so I found it worth asking for reviews to share your thoughts on this contribution. The earlier the changes are introduced, the lower the cost of their implementation. We also sometimes do not know all API features as they may not be widely promoted, but from your perspective they are important.  For example, all requests to Google API contain client info, which allows them to track API usage by a specific solution
https://github.com/apache/airflow/blob/d353f023ff8856c00b9f054526cb2e40ff0116ae/airflow/providers/google/common/hooks/base_google.py#L340-L352
Snowflake provides have a similar feature: https://github.com/apache/airflow/blob/d353f023ff8856c00b9f054526cb2e40ff0116ae/airflow/providers/snowflake/hooks/snowflake.py#L184 

On the other hand, it can also be a signal for you that there is new feature in the third project and you can promote it, e.g. by updating the documentation https://docs.getdbt.com/docs/running-a-dbt-project/running-dbt-in-production#using-airflow

Does it make sense to you?

**Comment by sungchun12** [comment]
@mik-laj thanks a bunch for the follow up. I'm in conversations internally with the dbt Labs engineers to specify a User-Agent in the headers of the dbt Cloud API requests.

I'll submit a pull request for the dbt Labs docs after this is merged!

@josh-fell I'll follow up with you personally on the above.

**Comment by potiuk** [comment]
Would you please rebase @josh-fell ?

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (sungchun12) [inline airflow/providers/dbt/cloud/hooks/dbt.py:88]: Add in a user agent to the header, and dynamically pass in the version of the airflow dbt Cloud provider to track usage on the dbt Cloud side. 

```suggestion
        request.headers["User-Agent"] = f"airflow-dbt-cloud-provider-v{X.Y.Z}"
```

**Comment by josh-fell** [comment]
> ... specify a User-Agent in the headers of the dbt Cloud API requests for tracking.

@sungchun12 @mik-laj This is implemented now if you'd like to take a look.  I also refactored the operator link slightly to not create ad hoc TaskInstances to align with #21285.

**Review by sungchun12 — APPROVED** [review]
Thanks for adding the change Josh. I verified it's tracking on the dbt Cloud side in our logs!

File: airflow/providers/dbt/cloud/hooks/dbt.py
Comment (sungchun12) [reply airflow/providers/dbt/cloud/hooks/dbt.py:88]: Note: the committed code is different from this suggestion!

Comment (sungchun12) [reply airflow/providers/dbt/cloud/hooks/dbt.py:88]: Note: the committed code is different from this suggestion!

**Comment by josh-fell** [comment]
@mik-laj Are there any other nuances/features I should try to incorporate in the provider? The last suggestion was a great one.

**Review by potiuk — APPROVED** [review]
I think it's a great DBT provider start. Just in time for Feb release.

**Comment by sungchun12** [comment]
@josh-fell amazing work, thanks for driving this to the finish line!

**Pull request merged**
